File: 2022/04/04.py
with open('input.txt', 'r') as file:
    pairs = file.read().splitlines()


def range_to_ints(range: str): return tuple(map(int, range.split('-')))


def ranges(elf1: str, elf2: str): return (
    *range_to_ints(elf1), *range_to_ints(elf2))


part1 = []
part2 = []

for pair in pairs:
    elf1, elf2 = pair.split(',')
    x1, x2, y1, y2 = ranges(elf1, elf2)
    part1.append(True if (x1 <= y1 and x2 >= y2)
                 or (x1 >= y1 and x2 <= y2) else False)
    part2.append(True if (x1 <= y1 and x2 >= y1)
                 or (x1 >= y1 and x1 <= y2) else False)

print(sum(part1), sum(part2))

File: 2023/scala3/src/main/scala/04/04.scala
package day4

@main
def main =
  val file =
    scala.io.Source.fromResource("04/input.txt").getLines().mkString("\n")
  val prepared = prep(file)
  println(part1(prepared))
  println(part2(prepared))

case class Card(winners: Array[Int], draw: Array[Int])
case class Copy(id: Int, bonuses: List[Int], var copies: Int = 1)

def toIntArr(line: String): Array[Int] =
  line.split(' ').map(_.toIntOption).flatten

def prep(input: String) =
  input
    .split('\n')
    .map: line =>
      val split = line.split('|')
      (
        split.headOption.map(toIntArr),
        split.lastOption.map(toIntArr)
      ) match {
        case (Some(winners), Some(draw)) => Some(Card(winners, draw))
        case _                           => None
      }
    .flatten

def wins(card: Card): Int =
  card.winners.count(card.draw.contains(_))

def score(card: Card): Int =
  val pow = wins(card) - 1
  if pow >= 0 then Math.pow(2, pow).toInt else 0

def part1(prep: Array[Card]) =
  prep.map(score).sum

def part2(prep: Array[Card]) =
  val bonuses = prep
    .map(wins)
    .zipWithIndex
    .map({ case (win, index) =>
      Copy(index + 1, List.range(index + 2, index + 2 + win))
    })
  for
    copy <- bonuses
    child <- copy.bonuses
  do
    bonuses.find(_.id == child) match {
      case Some(orig) => orig.copies += copy.copies
      case None       => None
    }
  bonuses.map(_.copies).sum
